src/views/Viewer.vue

```diff
@@ -555,9 +559,27 @@ export default {
 					displayName: t('viewer', 'View'),
 					mime: mime,
 					permissions: OC.PERMISSION_READ,
-					actionHandler: (name, { dir }) => {
+					actionHandler: (name, context) => {
+						const pushToHistory = ({ fileid }) => {
+							const params = OC.Util.History.parseUrlQuery()
+							const dir = params.dir
+							delete params.dir
+							delete params.fileid
+							params.openfile = fileid
+							const query = 'dir=' + OC.encodePath(dir) + '&' + OC.buildQueryString(params)
+							OC.Util.History.pushState(query)
+						}
 						// replace potential leading double slashes
-						OCA.Viewer.open(`${dir}/${name}`.replace(/^\/\//, '/'))
+						const path = `${context.dir}/${name}`.replace(/^\/\//, '/')
+						const oldQuery = location.search.replace(/^\?/, '')
+						const onClose = () => OC.Util.History.pushState(oldQuery)
+						if (!context.fileInfoModel && context.fileList) {
+							context.fileInfoModel = context.fileList.getModelForFile(name)
+						}
+						if (context.fileInfoModel) {
+							pushToHistory({ fileid: context.fileInfoModel.get('id') })
+						}
```

Comment:
Please put this whole logic into another file or service :)
This is a bit too messy.


Also, don't use `OC.encodePath`, or `OC.buildQueryString`.
We have dedicated libraries for this ;)
https://docs.nextcloud.com/server/latest/developer_manual/app/javascript-apis.html#npm-packages

Reply:
- [x] Move the logic into `src/services/FilesActionHandler`.
- [x] Replace `OC.encodePath`
- [ ] ~~Replace `OC.buildQueryString`~~ -  did not find a replacement. Is also not deprecated as of now.